File: ruby/circular-buffer/circular_buffer.rb
class CircularBuffer

  class BufferEmptyException < StandardError
  end
  class BufferFullException < StandardError
  end

  attr_reader :buffer, :head, :tail, :size

  def initialize(size)
    @size = size
    clear
  end

  def clear()
    @buffer = Array.new(size)
    @head = 0
    @tail = 0
  end

  def read
    raise BufferEmptyException if empty?
    value = buffer[(head + size) % size]
    buffer[(head + size) % size] = nil
    @head += 1
    value
  end

  def write(value)
    raise BufferFullException if full?
    write!(value)
  end

  def write!(value)
    @head += 1 if full?
    @buffer[(tail + size) % size] = value
    @tail += 1
  end

  private

  def full?
    head + size <= tail
  end

  def empty?
    head == tail
  end

end

File: javascript/circular-buffer/circular-buffer.js
var circularBuffer = function(max) {
  var buffer = [];

  return {
    clear: function() {
      buffer = [];
    },

    forceWrite: function(value) {
      if (buffer.length >= max) buffer.shift();
      if (value) buffer.push(value);
    },

    read: function() {
      if (buffer.length <= 0) throw bufferEmptyException();
      return buffer.shift();
    },

    write: function(value) {
      if (buffer.length >= max) throw bufferFullException();
      this.forceWrite(value);
    }
  };
};

var bufferEmptyException = function() {
  return {
    name: 'bufferEmptyException',
    message: 'cannot read empty buffer',
    toString: function() { return this.name + ": " + this.message; }
  };
};
bufferEmptyException.prototype = Object.create(Error.prototype);

var bufferFullException = function() {
  return {
    name: 'bufferFullException',
    message: 'cannot write to full buffer',
    toString: function() { return this.name + ": " + this.message; }
  };
};
bufferFullException.prototype = Object.create(Error.prototype);

module.exports = {
  circularBuffer: circularBuffer,
  bufferEmptyException: bufferEmptyException,
  bufferFullException: bufferFullException
};
